Nuvoton NCT6793D (and compatible) SMBus driver

CalcProgrammer1 opened the pull request:
This pull request implements a driver for the SMBus host controller present in several Nuvoton Super-IO chips including the NCT6793D.  This SMBus host is used for RGB LED control on several motherboards, including ASUS Intel Z270 and Z370 boards with ASUS Aura lighting.

This is my first time using the Pawn programming language or messing with PawnIO so I might need some help along the way.  So far I've just started the function to probe the chip and copied over some defines from the other versions of my driver.  It is not ready to merge yet but I wanted to get a pull request open to track progress and gather feedback.

I do have a test system that has this Super IO and RGB chip, ASUS PRIME Z270 PRO with Windows 10.

I have written an existing WinRing0-based driver in OpenRGB here:

https://gitlab.com/CalcProgrammer1/OpenRGB/-/blob/master/i2c_smbus/i2c_smbus_nct6775.cpp?ref_type=heads

I have also written an out-of-tree Linux driver here:

https://gitlab.com/CalcProgrammer1/i2c-nct6793-dkms/-/blob/master/src/i2c-nct6793.c?ref_type=heads

The Nuvoton NCT6793D datasheet is attached for reference, see Chapter 19, SMBUS MASTER INTERFACE.

[Nuvoton NCT6793D Datasheet.pdf](https://github.com/user-attachments/files/20683747/Nuvoton.NCT6793D.Datasheet.pdf)

namazso commented:
Looking pretty good so far! Once you're done with porting the Linux kmod features, you might want to look into https://github.com/namazso/PawnIO.Modules/pull/13 where the timing was perfected for piix4 and i801 compared to the loose "just sleep 250us" that Linux does.

CalcProgrammer1 commented:
Please see my comment about adjusting the ioctl interface for these SMBus modules to more closely match how Linux's ioctls (and our existing WinRing0 based Windows versions) function.

https://gitlab.com/CalcProgrammer1/OpenRGB/-/merge_requests/2885#note_2600407736

I would like to avoid having to duplicate the logic for all the different SMBus operations across each driver when we already have common code OpenRGB side to break down all the different SMBus calls into a single function call, which then just has to be undone and split back out into the individual ioctls for the current Pawn implementation.

CalcProgrammer1 commented:
Just refreshed my changes with the common API rework we've discussed in #15.  Still not complete or tested, but it does compile.  I need to flesh out the initial device probing and block operations.  I think I need to work on byte unpacking as well as the indices in the in/out arrays are 64-bit cells, not bytes, which are not accounted for properly yet.

CalcProgrammer1 commented:
```
[Info:] PawnIO initialized
[Info:] [ResourceManager] Registering I2C interface: i801 Device 8086:A2A3 Subsystem: 1043:872F
[ERROR:] Failed to load SmbusPIIX4.bin, PawnIO initialization aborted
[ERROR:] Failed to load SmbusPIIX4.bin, PawnIO initialization aborted
[Info:] PawnIO initialized
[Info:] [ResourceManager] Registering I2C interface: NCT6793 Device 0000:0000 Subsystem: 0000:0000
```

I got it to at least report its name, and it properly detects which Nuvoton chip I have so it is reading the SuperIO.  I have not tried to get it actually talking yet.

namazso commented on SmbusNCT6793.p:
```diff
@@ -0,0 +1,479 @@
+    out[1] = nuvoton_nct6793_smba;
+
+    // Not a PCI device, set PCI information to zero
+    out[2] = 0;
```
Vendor id 0000 is assigned to **Gammagraphx, Inc**. I think you're looking for VID:DID of FFFF:FFFF which is the no response value from bus master.

CalcProgrammer1 replied:
Good catch, I've updated it.

CalcProgrammer1 commented:
I've implemented my `set_byte_le` and `get_byte_le` functions to directly access bytewise data packed in 64-bit cell arrays and updated my NCT6793 driver to use them.  It detects the NCT6793 chip and can perform byte data read/write operations successfully, which is enough to detect that the ENE chip exists on my ASUS motherboard and set the mode, but block operations are used for setting colors which doesn't seem to work yet.  I've updated this branch and the pawnio_api_rework branch in OpenRGB with my latest changes.

namazso commented:
I've pushed an alternative version of those functions to main, so that they don't unnecessarily bypass bound checks if they were otherwise available:

<img width="853" height="800" alt="image" src="https://github.com/user-attachments/assets/46a6a8a7-0513-4a49-87ec-7e6d031c9194" />

As a general review, I'd much prefer if everywhere where possible static arrays sizes were used, as that ensures bounds checks are generated by the compiler.

CalcProgrammer1 commented:
Rebased and squashed.  I updated to using your new macros and made the arguments fixed size.  I was able to successfully set colors on my motherboard's ENE controller in direct mode, which, including detection, encompasses:

* Byte Data Read
* Byte Data Write
* Word Data Write
* Block Data Write

<img width="2560" height="1440" alt="image" src="https://github.com/user-attachments/assets/6f241a57-ec42-4aaf-be13-74b60fa42047" />

Before I unmark it as draft, I want to make sure block reads are working.  I don't think I've implemented this yet, but it should be able to read the colors out of the controller using a block read.

CalcProgrammer1 commented:
Actually looks like I did not implement read block in the Linux version of the driver.  ENE chips do not support block reads as far as we know so I have no way to test it.  It reads the colors out using byte or word reads.

CalcProgrammer1 commented:
Pushed again.  The datasheet for this controller is very vague so I hesitate to change the timing too much, but I did reduce the retry count based on some math.  It says the default frequency is 100KHz in the datasheet.  The data register acts as a 4-byte FIFO, so at most one continuous transfer consists of ADDR+RW, Index, and 4 bytes of data.  After this point, it waits until FIFO is empty and fills it with up to 4 more bytes.  Therefore the wait needs to be at most 6 bytes of time.  With the approximation of 10 bits per byte, 0.01ms per bit * 10 bits per byte * 6 bytes = 0.6ms wait per FIFO fill.  Allowing for roughly 4x that time reduces the wait retry count to 8 (250us per wait).

With this update I did not notice any issues detecting or controlling the ENE controller on my motherboard.

At this point I've unmarked draft as I've cleaned up the code and I think it's ready for final review.

hexawyz commented:
> ENE chips do not support block reads as far as we know so I have no way to test it.
You should try something like this: https://github.com/hexawyz/Exo/blob/master/src/Exo/Devices/Exo.Devices.Asus.Aura/AuraRamDriver.cs#L285
Works for me so I would assume it works in multiple cases 🙂

Logic:
If you observe the pattern of how the accesses work (which, except for the typo on the read 1 byte command, I actually got from your very old doc, so thanks 😉), it seems like the read and the writes are well separated.
Because `0x03` would be the write block command, one could assume that `0x83` could be a read block command of some sort (but how do you pass the parameter then?).
If you "just try it", it will return 3 bytes to you…
Then, `0x81` returns 1 byte, `0x82` returns 2 bytes, `0x83` returns 3 bytes… And it turns out that `0x84` does return 4 bytes. There is a upper limit as to how many bytes you can read at once, but up to 24 bytes at least works. (For me)

CalcProgrammer1 commented:
Interesting, I'll give it a try

On Tuesday, July 15, 2025, Fabien Barbier ***@***.***> wrote:
> <
https://ci3.googleusercontent.com/meips/ADKq_NbQ7MVmZ9cWNWDm-05BRykNMexKaqcnJ7qSmzblEKhrjMAbe27rFxlR9W-uhpZJGF-ODKbrME7euhmOlZ3o_43xzYUqDsbaQwnHdilyRw=s0-d-e1-ft#https://avatars.githubusercontent.com/u/8518235?s=20&v=4>hexawyz
left a comment (namazso/PawnIO.Modules#14)
>
> ENE chips do not support block reads as far as we know so I have no way
to test it.
>
> You should try something like this:
https://github.com/hexawyz/Exo/blob/master/src/Exo/Devices/Exo.Devices.Asus.Aura/AuraRamDriver.cs#L285
> Works for me so I would assume it works in multiple cases 🙂
>
> Logic:
> If you observe the pattern of how the accesses work (which, except for
the typo on the read 1 byte command, I actually got from your very old doc,
so thanks 😉), it seems like the read and the writes are well separated.
> Because 0x03 would be the write block command, one could assume that 0x83
could be a read block command of some sort (but how do you pass the
parameter then?).
> If you "just try it", it will return 3 bytes to you…
> Then, 0x81 returns 1 byte, 0x82 returns 2 bytes, 0x83 returns 3 bytes…
And it turns out that 0x84 does return 4 bytes. There is a upper limit as
to how many bytes you can read at once, but up to 24 bytes at least works.
(For me)
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.<
https://ci3.googleusercontent.com/meips/ADKq_Nbe0J7VvxkZD8IuR1Tuc83sa-L2ygzTG9Qi5WrtZpDtEWdXfbzg2X6DFuraYjwIQWXWvRax4E03g-qvW3ONQQfzlddrdGu9JD_ppTVSbcZmRMBZUFyKXFY2uYgYcCzSIIf5D23YNoIu2AfdN5OUhg8me9kjkWnlVvpioLZwnQOqoLZ8we_7E-QGoDpcrsnz69bnERz9ExC73OJJ-pBFsBa5g9Kzh6VjXRWxlTAYWT_w2niiah8aWCQ=s0-d-e1-ft#https://github.com/notifications/beacon/AAIAY7ET47O7GGVPKCPIBBD3IVXYLA5CNFSM6AAAAAB7BK4GL6WGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVXKTBXU.gif>Message
ID: ***@***.***>

CalcProgrammer1 commented on SmbusNCT6793.p:
```diff
@@ -0,0 +1,445 @@
+/// @note The NCT6793 controller use a fixed clock frequency.
+/// @return An NTSTATUS
+forward NTSTATUS:ioctl_clock_freq(in[], in_size, out[], out_size);
+public NTSTATUS:ioctl_clock_freq(in[], in_size, out[], out_size)
```
So the datasheet says that register offset 7 is the SCLFREQ clock register, but reading register offset 7 seems to always return 0.  I noticed there was a gap and register address 6 isn't used so I tried reading 6 instead and it read the correct default value (0b00000111, which is 99KHz).

<img width="901" height="824" alt="image" src="https://github.com/user-attachments/assets/870ba2a4-90b3-4e5b-a82e-820b10f5941d" />

I think the datasheet is wrong here.  I think offset 6 is the clock frequency register.  I'm going to do more testing.

I haven't found any other references to this controller or other open source implementations so I really have no way to be sure.

Steve-Tech replied:
> I think the datasheet is wrong here. I think offset 6 is the clock frequency register.

To add to your suspicions, the register map above has it at offset 6, so I'd say that's correct!

<img width="827" height="823" alt="image" src="https://github.com/user-attachments/assets/cfbb61b3-f29b-46e1-908d-e13fe27f9593" />

CalcProgrammer1 replied:
I can set and read back register offset 6 correctly, but I tried value 0 and I tried value 15 and couldn't observe any difference interacting with the ENE RGB chip, but that doesn't mean anything.  I think I'll have to get out the oscilloscope to verify this register.

CalcProgrammer1 replied:
Based on that summary saying it's offset 6 and the fact that it read the correct default I think I'm confident in my changes.  It can now read but not set the clock.

CalcProgrammer1 replied:
Rebased on main, updated to be able to set frequency as well.

CalcProgrammer1 replied:
It looks like soft reset might be resetting the set clock, as I can change it and immediately read back and it's different, but if I restart OpenRGB (reload the driver) it goes back to default.  I'm only debug_printing at module load time so it could be changing immediately on the first access.

CalcProgrammer1 replied:
Yeah this is exactly what's happening, printing out the clock after the reset call it's always 99KHz.

CalcProgrammer1 replied:
Ok, I've updated it to have a variable, the initial setting of the clock register is read in during initialization and is written out after reset.  It looks like it holds its value now.

CalcProgrammer1 replied:
I think it's ready to merge now.  The only thing not implemented is block read which isn't needed for the OpenRGB use case, but looks like it may be possible to test with ENE.  I think it's fine to merge without this capability now, but I'm open to adding it later in a future MR.

hexawyz replied:
The use case for block reads is a bit niche, as you would basically only need that during startup, but it allows you to reduce the number of SMBus I/O operations by reading multiple setting bytes at once, so it would still be a nice improvement. (e.g. a single call for reading 24 color bytes instead of 24 calls)
I have not yet found out how to best integrate PawnIO SMBus support in my own use case (purely architectural concerns) but I would definitely make use of it 😅

CalcProgrammer1 commented:
For this, since it's an auxiliary SMBus only used for the onboard ENE chip,
I would be probing the ENE chip directly.  I don't think this SMBus pins
are exposed on any pin headers or connectors.

On Mon, Jul 28, 2025, 1:08 AM Stephen Horvath ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In SmbusNCT6793.p
> <https://github.com/namazso/PawnIO.Modules/pull/14#discussion_r2234803721>
> :
>
> > +    /* Not a PCI device, set PCI information to invalid (FFFF:FFFF, FFFF:FFFF) */
> +    out[2] = 0xffffffffffffffff;
> +
> +    return STATUS_SUCCESS;
> +}
> +
> +/// Set the SMBus clock frequency.
> +///
> +/// @param in [0] = Frequency in Hz or -1 for no change
> +/// @param in_size Must be 1
> +/// @param out [0] = Previous frequency in Hz
> +/// @param out_size Must be 1
> +/// @note The NCT6793 controller use a fixed clock frequency.
> +/// @return An NTSTATUS
> +forward NTSTATUS:ioctl_clock_freq(in[], in_size, out[], out_size);
> +public NTSTATUS:ioctl_clock_freq(in[], in_size, out[], out_size)
>
> I think I'll have to get out the oscilloscope to verify this register.
>
> Just curious, what is the easiest way to physically access the SMBus on
> ASUS boards? Would you just be clipping/soldering the probes to one of the
> DIMM pins, or is there a header like COM_DEBUG?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/namazso/PawnIO.Modules/pull/14#discussion_r2234803721>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIAY7F6YL6PFECOJHTAHWT3KW46RAVCNFSM6AAAAAB7BK4GL6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTANRQGU2DGOJQGA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
